File: 454. 4Sum II.cpp
class Solution {

public:

    int fourSumCount(vector<int>& A, vector<int>& B, vector<int>& C, vector<int>& D) {

        unordered_map<int, int> hash;

        int cnt = 0;

        if(A.size() * B.size() * C.size() * D.size()==0) return cnt;

        for(int i=0; i< A.size(); i++){

            for(int j=0; j<B.size(); j++){

                hash[A[i]+B[j]]++;

            }

        }

        for(int i=0; i< C.size(); i++){

            for(int j=0; j<D.size(); j++){

                cnt += hash[-C[i]-D[j]];

            }

        }

        return cnt;

    }

};

File: 454. 4Sum II.py
class Solution(object):

    def fourSumCount(self, A, B, C, D):

        """

        :type A: List[int]

        :type B: List[int]

        :type C: List[int]

        :type D: List[int]

        :rtype: int

        """

        AB = collections.defaultdict(int)

        ans = 0

        for a in A:

            for b in B:

                AB[a+b] += 1

        for c in C:

            for d in D:

                ans += AB[-c-d]

        return ans
